### PR TITLE
fix: serialization of non-model dict objects (fixes #155 #722 #1126)

### LINKED ISSUE #155: Support custom object keys
Hey, thanks for your work on this project. I was investigating integrating it with our EmberJS app but I ran into an issue. We have an `Employee` model that has a normal integer primary key, but also a `username` attribute. I'd like to only expose the `username` as the primary key, using DRF this is simple:

```python
class EmployeeSerializer(serializers.ModelSerializer):
    id = serializers.ReadOnlyField(source="user.username")
    ...
```

However in `build_json_resource_obj` it just uses instance.pk:

```python
def build_json_resource_obj(fields, resource, resource_instance, resource_name):
    resource_data = [
        ('type', resource_name),
        ('id', encoding.force_text(resource_instance.pk) if resource_instance else None),
        ...
```

Can you use the serializers ID field if available, and fallback to the instances PK?

### PR DESCRIPTION
**Pull request by axieum**
Fixes #155 #722 #1126

## Description of the Change

Redirect logic used to extract the resource identifier from a serializer instance to a utility method. This utility method handles the edge-case where there is no model backing the serializer resulting in a "`dict` object has no attribute `pk`".

## Checklist

- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [ ] documentation updated
- [ ] `CHANGELOG.md` updated (only for user relevant changes)
- [ ] author name in `AUTHORS`

### CODE REVIEW

**Comment by sliverc** [comment]
Thanks for looking into our oldest open issue. It is great to tackle this.

I think it is a great idea to allow a DJA user to overwrite an `id` by configuring an `id` field on the serializer, as this is most intuitive. I think your code needs to be adjusted though, as in its current states it expects an `id` on the instance, and not on the serialized data.

A check would need to do the following:
1. check if there is an `id` in the serialized data (in the method `build_json_resource_obj` the serialized data is called `resource`).
2. if not fall back to the pk on the resource instance

That should be enough in my opinion and no more if/else are needed. It will also be important to document this behavior in our documentation, that setting an `id` field on the serializer is the way to have a custom `id` or when using generic serializers. To fully resolve #155 it also needs to be possible to configure a custom `pk` for a relation. When digging into the code, this is actually already possible by setting [pk_field](https://github.com/django-json-api/django-rest-framework-json-api/blob/main/rest_framework_json_api/relations.py#L251) on the `ResourceRelatedField`, so we would only need to document and add a test.

You think you could work on those things?

**Comment by sliverc** [comment]
@axieum Do you think you can still work on this PR (open issues see comment above)? I very much like the implementation and would love to get it integrated.

**Comment by axieum** [comment]
Hi, sorry about that @sliverc - got heavily distracted elsewhere.

I've gone ahead and quickly added some commits to do what you proposed, first it'll try the `resource` and then the `resource_instance`. In both scenarios, the `resource` (and `resource_instance`) may be a `dict` or `object`.

I couldn't think of how to write a test for the case where `resource` is preferred over `resource_instance`, could you assist here please?

Thanks,
Jonathan.

**Comment by sliverc** [comment]
```python
def get_resource_id(resource_instance, resource):
   if resource and 'id' in resource:
      return force_str(resource['id'])
   return resource_instance and force_str(resource_instance.pk) or None
```

As we need to add a utility method anyway to get the resource id, force_str, checking for None etc. should all be in there. Above is how the method could look like (not tested, but as an idea). As passing on `resource_instance` and `resource` separately, it is actually much simpler.

For writing a test, you need to write a test where the serializer has a field called id. Easiest would then be to add a method get_id(self) to the serializer which returns a different id to see whether that id is actually returned. Then test it as you already do in the `test_views.py` file.

Hope this makes things clearer.

**Comment by axieum** [comment]
I think we're going to need to explicitly add code to check for and execute a `get_id` method for that piece to work.

File: rest_framework_json_api/utils.py [hunk 1]
```diff
@@ -304,6 +304,21 @@ def get_resource_type_from_serializer(serializer):
     )
 
 
+def get_resource_id_from_instance(resource_instance, resource):
+    """Returns the resource identifier for a given instance (`id` takes priority over `pk`)."""
+    if resource and "id" in resource:
+        return resource["id"] and encoding.force_str(resource["id"]) or None
+    if resource_instance:
+        if hasattr(resource_instance, "get_id") and callable(resource_instance.get_id):
```
Comment (sliverc) [inline rest_framework_json_api/utils.py:312]: What is the use case to have a `get_id` method on the resource instance resp. model? The model is the database representation, whereas the `id` is the JSON:API spec representation. So if the `id` needs to be different from the database it should be defined on the serializer by adding a `id` field. That should be enough as far as I see it.

File: rest_framework_json_api/utils.py [hunk 2]
```diff
@@ -304,6 +304,21 @@ def get_resource_type_from_serializer(serializer):
     )
 
 
+def get_resource_id_from_instance(resource_instance, resource):
```
Comment (sliverc) [inline rest_framework_json_api/utils.py:307]: I would rename it now as it is not retrieving more than just an instance to `get_resource_id`.

**Review by sliverc — CHANGES_REQUESTED** [review]
Thanks for continue working on this. See my inline comments + the tests do not seem to pass, therefore haven't reviewed those in detail yet.

**Comment by sliverc** [comment]
This is a very important feature and as we are soon releasing a new version I would like to get this ready. So I am currently working on it to be able to merge it before our next release.

**Review by sliverc — APPROVED** [review]

**Comment by sliverc** [comment]
I now added some documentation as well. For this feature to be complete, it is also necessary to be able to overwrite resource id for resource related fields. Added this now as well. So this is ready to merge. In case there are any comments of anything I missed, let me know.
